quesma/queryparser/pancake_json_rendering.go

```diff
@@ -209,58 +233,71 @@ func (p *pancakeJSONRenderer) layerToJSON(remainingLayers []*pancakeModelLayer,
 			return result, nil
 		}
 
-		if len(remainingLayers) > 1 { // Add subAggregation
-			if bucketArrRaw, ok := buckets["buckets"]; ok {
-				bucketArr := bucketArrRaw.([]model.JsonMap)
+		hasSubaggregations := len(remainingLayers) > 1
+		if hasSubaggregations {
+			nextLayer := remainingLayers[1]
+			pipelineBucketsPerAggregation := p.pipelinesProcessor.calculateThisLayerBucketPipelines(layer, nextLayer, bucketRows, rows, rowIndexes)
 
-				if len(bucketArr) == len(subAggrRows) {
+			// Add subAggregations (both normal and pipeline)
+			bucketArrRaw, ok := buckets["buckets"]
+			if !ok {
+				return nil, fmt.Errorf("no buckets key in bucket json, layer: %s", layer.nextBucketAggregation.name)
+			}
 
-					// Simple case, we merge bucketArr[i] with subAggrRows[i] (if lengths are equal, keys must be equal => it's fine to not check them at all)
-					for i, bucket := range bucketArr {
-						if docCount, ok := bucket["doc_count"]; ok {
-							if fmt.Sprintf("%v", docCount) == "0" { // Not sure, but it does the trick
-								continue
-							}
-						}
-						// TODO: Maybe add model.KeyAddedByQuesma if there are more than one pancake
-						subAggr, err := p.layerToJSON(remainingLayers[1:], subAggrRows[i])
+			bucketArr := bucketArrRaw.([]model.JsonMap)
+
+			if len(bucketArr) == len(subAggrRows) {
+				// Simple case, we merge bucketArr[i] with subAggrRows[i] (if lengths are equal, keys must be equal => it's fine to not check them at all)
+				for i, bucket := range bucketArr {
+					for pipelineAggrName, pipelineAggrBuckets := range pipelineBucketsPerAggregation {
+						bucketArr[i][pipelineAggrName] = model.JsonMap{"value": pipelineAggrBuckets[i].LastColValue()}
+					}
+
+					if docCount, ok := bucket["doc_count"]; ok && fmt.Sprintf("%v", docCount) == "0" {
+						// Not sure, but it does the trick.
+						continue
+					}
+
+					// TODO: Maybe add model.KeyAddedByQuesma if there are more than one pancake
+					subAggr, err := p.layerToJSON(remainingLayers[1:], subAggrRows[i])
+					if err != nil {
+						return nil, err
+					}
+					bucketArr[i] = util.MergeMaps(p.ctx, bucket, subAggr, model.KeyAddedByQuesma)
+				}
+			} else {
+				// A bit harder case. Observation: len(bucketArr) > len(subAggrRows) and set(subAggrRows' keys) is a subset of set(bucketArr's keys)
+				// So if bucket[i]'s key corresponds to subAggr[subAggrIdx]'s key, we merge them.
+				// If not, we just keep bucket[i] (i++, subAggrIdx stays the same)
+				subAggrIdx := 0
+				for i, bucket := range bucketArr {
+					for pipelineAggrName, pipelineAggrBuckets := range pipelineBucketsPerAggregation {
+						bucketArr[i][pipelineAggrName] = model.JsonMap{"value": pipelineAggrBuckets[i].LastColValue()}
```

Comment:
I don't like it here. Got different idea.

Reply:
Let's talk tomorrow.